--- aaa/src/src.pro	2013-05-18 22:30:23.978572868 +0400
+++ bbb/src/src.pro	2013-05-18 22:30:12.642572017 +0400
@@ -195,11 +195,6 @@ isEmpty(gconf_config_source) {
     gconf_config_source = $$system(gconftool-2 --get-default-source)
 }
 
-QMAKE_EXTRA_TARGETS += register_schemas
-register_schemas.target = register_schemas
-register_schemas.commands += GCONF_CONFIG_SOURCE=$$gconf_config_source $$gconftool --makefile-install-rule $$install_schemas.files
-install_schemas.depends += register_schemas
-
 # Check targets
 QMAKE_EXTRA_TARGETS += check-xml
 check-xml.target = check-xml
