--- hildon-desktop-clean/src/launcher/hd-launcher-editor.c	2012-07-13 21:14:41.776453670 +0400
+++ hildon-desktop/src/launcher/hd-launcher-editor.c	2012-07-13 21:16:29.186453039 +0400
@@ -400,8 +400,12 @@
                                                      G_TYPE_STRING);
 
   /* and the icon view. */
+#ifdef MAEMO_GTK
   priv->icon_view = hildon_gtk_icon_view_new_with_model (HILDON_UI_MODE_EDIT,
                                                          priv->model);
+#else
+  priv->icon_view = gtk_icon_view_new_with_model (priv->model);
+#endif
 
   renderer = gtk_cell_renderer_pixbuf_new ();
   g_object_set (G_OBJECT (renderer),
@@ -502,7 +506,7 @@
 {
   gtk_widget_realize (window);
 
-  GdkWindow *gdk_window = window->window;
+  GdkWindow *gdk_window = gtk_widget_get_window (window);
   GdkDisplay *gdk_display = gdk_display_get_default ();
   Display *display = GDK_DISPLAY_XDISPLAY (gdk_display);
   Window xwindow = xwindow = GDK_WINDOW_XID (gdk_window);
--- hildon-desktop-clean/src/main.c	2012-07-13 18:47:32.000000000 +0000
+++ hildon-desktop/src/main.c	2012-07-13 18:45:34.000000000 +0000
@@ -605,7 +605,7 @@
 
 static void hd_fpu_set_mode(OSSO_FPU_MODE mode)
 {
-#ifdef __arm__
+#if  defined(__arm__) && defined(__VFP_FP__) && !defined(__SOFTFP__)
   if (OSSO_FPU_FAST == mode)
   {
     int tmp;
