diff -ur clean/matchbox/core/mb-window-manager.c work/matchbox/core/mb-window-manager.c
--- clean/matchbox/core/mb-window-manager.c	2012-07-13 12:52:31.609708675 +0400
+++ work/matchbox/core/mb-window-manager.c	2012-07-13 12:54:19.434703198 +0400
@@ -386,7 +386,7 @@
   MBWindowManager *wm = (MBWindowManager*)userdata;
 
   mb_wm_keys_press (wm,
-		    XKeycodeToKeysym(wm->xdpy, xev->keycode, 0),
+		    XkbKeycodeToKeysym(wm->xdpy, xev->keycode, 0, 0),
 		    xev->state);
 
   return True;
diff -ur clean/matchbox/core/mb-wm-keys.c work/matchbox/core/mb-wm-keys.c
--- clean/matchbox/core/mb-wm-keys.c	2012-07-13 12:52:31.611708675 +0400
+++ work/matchbox/core/mb-wm-keys.c	2012-07-13 12:53:42.898705072 +0400
@@ -33,7 +33,7 @@
 
   for (keycode = min_kc; keycode <= max_kc; keycode++)
     for (col = 0;
-	 (k = XKeycodeToKeysym (wm->xdpy, keycode, col)) != NoSymbol;
+	 (k = XkbKeycodeToKeysym (wm->xdpy, keycode, col, 0)) != NoSymbol;
 	 col++)
       if (k == keysym && col == 1)
 	return True;
@@ -352,7 +352,7 @@
 	for (col = 0; col < 4; col += 2)
 	  {
 	    KeyCode code = mod_map->modifiermap[mod_idx * kpm + mod_key];
-	    KeySym sym = (code ? XKeycodeToKeysym(wm->xdpy, code, col) : 0);
+	    KeySym sym = (code ? XkbKeycodeToKeysym(wm->xdpy, code, col, 0) : 0);
 
 	    if (sym == last_sym) continue;
 	    last_sym = sym;
