--- /dev/null	2011-12-08 14:00:21.159999965 +0400
+++ NuxCore/nullptr.h	2011-12-09 20:52:05.998393984 +0400
@@ -0,0 +1,12 @@
+const                        // this is a const object...
+class {
+public:
+  template<class T>          // convertible to any type
+    operator T*() const      // of null non-member
+    { return 0; }            // pointer...
+  template<class C, class T> // or any type of null
+    operator T C::*() const  // member pointer...
+    { return 0; }
+private:
+  void operator&() const;    // whose address can't be taken
+} nullptr = {};              // and whose name is nullptr
--- NuxCore/NuxCore.h.orig	2011-12-09 21:06:08.998389032 +0400
+++ NuxCore/NuxCore.h	2011-12-09 20:58:39.968391670 +0400
@@ -22,6 +22,8 @@
 #ifndef NUXCORE_H
 #define NUXCORE_H
 
+#include "nullptr.h"
+
 #include "System.h"
 
 #include <cassert>  // (assert.h)

